test/parallel/test-assert-async.js

```diff
@@ -0,0 +1,96 @@
+'use strict';
+const common = require('../common');
+const assert = require('assert');
+const {promisify} = require('util');
+const wait = promisify(setTimeout);
+
+/* eslint-disable prefer-common-expectserror, no-restricted-properties */
+
+// Test assert.rejects() and assert.doesNotReject() by checking their
+// expected output and by verifying that they do not work sync
+
+assert.doesNotThrow(() => {
+  assert.rejects(
+    () => assert.fail(),
+    common.expectsError({
+      code: 'ERR_ASSERTION',
+      type: assert.AssertionError,
+      message: 'Failed'
+    })
+  )
+});
+
+assert.doesNotThrow(() => assert.doesNotReject(() => {}))
+
+{
+  let promise;
+  let threw = false;
+  try {
+    promise = assert.rejects(async () => {
+      await wait(10);
+      assert.fail();
+    }, common.expectsError({
+      code: 'ERR_ASSERTION',
+      type: assert.AssertionError,
+      message: 'Failed'
+    }));
+  } catch (err) {
+    threw = true;
+  }
+  assert.strictEqual(threw, false);
+  assert.doesNotReject(() => promise);
+}
+
+{
+  let threw = false;
+  let promise;
+  try {
+    promise = assert.doesNotReject(async () => {
+      await wait(10);
+      throw Error();
+    });
+  } catch (err) {
+    threw = true;
+  }
+  assert.strictEqual(threw, false);
```

Comment:
After thinking about this again, the try catch and the threw is not necessary here at all. If the promise throws the test would fail right away. `threw` is only necessary for tests that check for `threw === true`. So the tests can be much smaller (I know I originally suggested this pattern).